numcodecs/gzip.py

```diff
@@ -38,24 +38,30 @@ def encode(self, buf):
                             mode='wb',
                             compresslevel=self.level) as compressor:
             compressor.write(buf)
-        compressed = compressed.getvalue()
 
-        return compressed
+        try:
+            compressed = compressed.getbuffer()
+        except AttributeError:  # pragma: py3 no cover
+            compressed = compressed.getvalue()
+
+        return ensure_ndarray(compressed)
 
     # noinspection PyMethodMayBeStatic
     def decode(self, buf, out=None):
 
         # normalise inputs
         buf = ensure_contiguous_ndarray(buf)
-        if out is not None:
-            out = ensure_contiguous_ndarray(out)
 
         # do decompression
         buf = io.BytesIO(buf)
         with _gzip.GzipFile(fileobj=buf, mode='rb') as decompressor:
-            decompressed = decompressor.read()
+            if out is not None:
+                decompressed = ensure_contiguous_ndarray(out)
+                decompressor.readinto(decompressed)
```

Comment:
Should we check that the buffer was exhausted here? If so, what should we raise if that does not happen?